wd_util.c

```diff
@@ -1092,44 +1109,183 @@ static void wd_uninit_resource(struct wd_env_config *config)
+
+	snprintf(var_tbl[0].def_val, MAX_STR_LEN, "%s%u%c%u",
+		 type_tbl, attr->num, '@', attr->node);
+
+	return 0;
+}
+
+static void wd_alg_table_uninit(struct wd_env_config *config)
+{
+	int i;
+
+	if (!config->disable_env)
+		return;
+
+	for (i = 0; i < config->table_size - 1; i++)
+		free(config->table[i].def_val);
+
+	free((struct wd_config_variable *)config->table);
+}
+
 int wd_alg_env_init(struct wd_env_config *env_config,
 		    const struct wd_config_variable *table,
 		    const struct wd_alg_ops *ops,
-		    __u32 table_size)
+		    __u32 table_size,
+		    struct wd_ctx_attr *ctx_attr)
 {
 	int ret;
 
-	if (!env_config || !table || !table_size || !ops)
-		return -WD_EINVAL;
+	env_config->op_type_num = ops->op_type_num;
+	env_config->table_size = table_size;
 
-	ret = wd_alloc_numa(env_config);
+	ret = wd_alg_table_init(table, table_size,
+				ctx_attr, env_config);
 	if (ret)
 		return ret;
 
-	env_config->op_type_num = ops->op_type_num;
-	env_config->table = table;
-	env_config->table_size = table_size;
+	ret = wd_alloc_numa(env_config, ops);
+	if (ret)
+		goto table_uninit;
 
 	ret = wd_parse_env(env_config);
-	if (ret) {
-		wd_free_numa(env_config);
-		return ret;
-	}
+	if (ret)
+		goto free_numa;
 
 	ret = wd_init_resource(env_config, ops);
-	if (ret) {
-		wd_free_env(env_config);
-		wd_free_numa(env_config);
-		return ret;
-	}
+	if (ret)
+		goto free_env;
+
+	/* Use alg_uninit as a sign of initialization complete */
+	env_config->alg_uninit = ops->alg_uninit;
 
 	return 0;
+
+free_env:
+	wd_free_env(env_config);
+free_numa:
+	wd_free_numa(env_config);
+table_uninit:
+	wd_alg_table_uninit(env_config);
+	return ret;
 }
 
 void wd_alg_env_uninit(struct wd_env_config *env_config)
 {
+	/* Check whether the initialization is complete */
+	if (!env_config->alg_uninit)
+		return;
+
 	wd_uninit_resource(env_config);
 	wd_free_env(env_config);
+	wd_free_numa(env_config);
+	wd_alg_table_uninit(env_config);
+}
+
+int wd_alg_get_env_param(struct wd_env_config *env_config,
+			 struct wd_ctx_attr attr,
+			 __u32 *num, __u8 *is_enable)
+{
+	struct wd_env_config_per_numa *config_numa;
+
+	if (!num || !is_enable) {
+		WD_ERR("input parameter num or is_enable is NULL!\n");
+		return -WD_EINVAL;
+	}
+
+	*is_enable = env_config->enable_internal_poll;
+
+	config_numa = wd_get_config_numa(env_config, attr.node);
+	if (!config_numa) {
+		WD_ERR("%s got wrong numa node: %u!\n",
+				__func__, attr.node);
+		return -WD_EINVAL;
+	}
+
+	*num = config_numa->ctx_table[attr.mode][attr.type].size;
+
+	return 0;
+}
+
+int wd_set_ctx_attr(struct wd_ctx_attr *ctx_attr,
+		     __u32 node, __u32 type, __u8 mode, __u32 num)
+{
+	if (mode >= CTX_MODE_MAX) {
+		WD_ERR("wrong ctx mode(%u))!\n", mode);
+		return -WD_EINVAL;
+	}
+
+	ctx_attr->node = node;
+	ctx_attr->mode = mode;
+	ctx_attr->num = num;
+	/* If type is CTX_TYPE_INVALID, we need update it to 0. */
+	ctx_attr->type = (type == CTX_TYPE_INVALID) ? 0 : type;
```

Comment:
If mode parameter is invalid, assignment isn't executed. Now type parameter is invalid too, it seems that we should skip the assignment.

Reply:
type is set to 0, so we can use ctx_table[mode][type] also for cipher，rsa etc